mat64/dense_arithmetic.go

```diff
@@ -434,6 +434,168 @@ func (m *Dense) Mul(a, b Matrix) {
 	*m = w
 }
 
+func (m *Dense) MulTrans(a Matrix, aTrans blas.Transpose, b Matrix, bTrans blas.Transpose) {
+	ar, ac := a.Dims()
+	if aTrans != blas.NoTrans {
+		ar, ac = ac, ar
+	}
+
+	br, bc := b.Dims()
+	if bTrans != blas.NoTrans {
+		br, bc = bc, br
+	}
+
+	if ac != br {
+		panic(ErrShape)
+	}
+
+	var w Dense
+	if m != a && m != b {
+		w = *m
+	}
+	if w.isZero() {
+		w.mat = RawMatrix{
+			Rows:   ar,
+			Cols:   bc,
+			Stride: bc,
+			Data:   use(w.mat.Data, ar*bc),
+		}
+	} else if ar != w.mat.Rows || bc != w.mat.Cols {
+		panic(ErrShape)
+	}
+
+	//TODO(jonlawlor): if we are performing a * a' or a' * a, then make a call to
+	// blas.Dsyrk instead of blas.Dgemm.
+
+	if a, ok := a.(RawMatrixer); ok {
+		if b, ok := b.(RawMatrixer); ok {
+			amat, bmat := a.RawMatrix(), b.RawMatrix()
+			if blasEngine == nil {
+				panic(ErrNoEngine)
+			}
+			blasEngine.Dgemm(
+				aTrans, bTrans,
+				ar, bc, ac,
+				1.,
+				amat.Data, amat.Stride,
+				bmat.Data, bmat.Stride,
+				0.,
+				w.mat.Data, w.mat.Stride)
+			*m = w
+			return
+		}
+	}
+
+	if a, ok := a.(Vectorer); ok {
+		if b, ok := b.(Vectorer); ok {
+			row := make([]float64, ac)
+			col := make([]float64, br)
+			if blasEngine == nil {
+				panic(ErrNoEngine)
+			}
+			if aTrans != blas.NoTrans {
+				if bTrans != blas.NoTrans {
+					for r := 0; r < ar; r++ {
```

Comment:
The trouble is that would require a type assert to access the mat field, and we don't currently have a type that implements Vectorer but doesn't implement RawMatrixer.